pdns/protozero.hh

```diff
@@ -0,0 +1,254 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#pragma once
+
+#include <protozero/pbf_writer.hpp>
+#include <string>
+
+#include "config.h"
+#include "iputils.hh"
+#include "filterpo.hh"
+#include "gettime.hh"
+#include "uuid-utils.hh"
+
+namespace pdns {
+  namespace ProtoZero {
+    class Message {
+    public:
+      // Start a new messagebug, constaining separate data for the response part
+      Message(std::string::size_type sz1, std::string::size_type sz2) : d_message{d_msgbuf}, d_response{d_rspbuf}
+      {
+        // This is extra space in addition to what's already there
+        // Different from what string.reserve() does
+        d_message.reserve(sz1);
+        d_response.reserve(sz2);
+      }
+
+      // Construct a Message with (partially) constructed content
+      Message(const std::string& buf1, const std::string& buf2, std::string::size_type sz1, std::string::size_type sz2) :
+        d_msgbuf{buf1}, d_rspbuf{buf2}, d_message{d_msgbuf}, d_response{d_rspbuf}
+      {
+        // We expect to grow the buffers, in the end the d_message will contains the (grown) d_response
+        // This is extra space in addition to what's already there
+        // Different from what string.reserve() does
+        d_message.reserve(sz1 + d_rspbuf.length() + sz2);
+        d_response.reserve(sz2);
+      }
```

Comment:
Perhaps we could mark the copy constructor, copy assignment operator, move constructor and move assignment operator as `delete`?

Reply:
That might not play too well with the `boost::optional` suggestion, though, we would need to check.

Reply:
I tried to use `boost::optional` again, but failed to do so. My guess is that because `pbf_writer` contains a pointer to the associated string, something goes wrong somewhere, even though the `pbf_writer` object itself *is* movable... But I did not figure out yet how to fix that.